### PR TITLE
bpo-38852: Set thread stack size to 8 Mb for debug builds on android platforms

### PR DESCRIPTION
**Pull request by xdegaye**
https://bugs.python.org/issue38852

### CODE REVIEW

File: Python/thread_pthread.h [hunk 1]
```diff
@@ -51,6 +51,12 @@
 #undef  THREAD_STACK_SIZE
 #define THREAD_STACK_SIZE       0x200000
 #endif
+#if defined(__ANDROID__) && defined(THREAD_STACK_SIZE) && THREAD_STACK_SIZE == 0
+#   ifdef Py_DEBUG
```
Comment (vstinner) [inline Python/thread_pthread.h:59]: Please add a comment explaining why this is needed. I understand the Python compiled in debug mode uses more stack and that test_threading.test_recursion_limit() requires to be able to call up to 1000 Python frames in a thread without crashing. I propose the comment:

```
/* bpo-38852: test_threading.test_recursion_limit() checks that 1000 recursive Python calls
   (default recursion limit) doesn't crash, but raise a regular RecursionError exception.
   In debug mode, Python function calls allocates more memory on the stack, so use a
   stack of 8 MiB. */
```

**Comment by vstinner** [comment]
Thanks for the change. I don't understand why debug build uses 7x more stack memory, but I trust your tests https://bugs.python.org/issue38852#msg357266 :-)

@xdegaye: Do you want to backport the change to Python 3.7 and/or 3.8?

**Comment by xdegaye** [comment]
@vstinner

> I don't understand why debug build uses 7x more stack memory

In the initial post of issue https://bugs.python.org/issue18049, Ronald considers that it is the ``-O0`` optimisation level in debug builds that causes this increase in stack memory usage. The commits that fix this issue do not use the #ifdef PY_DEBUG conditional suggested by Stéphane in the third post because, as I understand it, it is considered that "users could also build with CFLAGS set to -O0 without using --with-pydebug". 

BTW in both android and OSX cases it seems that it is **clang** with debug builds that causes the problem. See issue https://bugs.python.org/issue14184, the initial change in stack size on OSX.

> Do you want to backport the change to Python 3.7 and/or 3.8?

Since only debug builds are concerned with this change and android is not officially supported, I think backports are not necessary.

**Comment by vstinner** [comment]
> In the initial post of issue https://bugs.python.org/issue18049, Ronald considers that it is the -O0 optimisation level in debug builds that causes this increase in stack memory usage.

Ah yes, that makes sense.

> Since only debug builds are concerned with this change and android is not officially supported, I think backports are not necessary.

Ok, I'm fine with that.

**Pull request merged**